app/controllers/graphql_controller.rb

```diff
@@ -1,13 +1,32 @@
 class GraphqlController < ApplicationController
+  before_action :authenticate
+
   def execute
     query_string = params[:query].to_s
     variables = ensure_hash(params[:variables])
-    result = ::Graph::Schema.execute(query_string, variables: variables)
+    context = {
+      user: @user
+    }
+
+    result = Graph::Schema.execute(query_string, variables: variables, context: context)
     render json: result
   end
 
   private
 
+  def authenticate
+    @user = authenticate_with_http_basic { |username, password|
+      user = User.where(username: username).first
+
+      return render plain: 'Invalid username', status: :unauthorized unless user
```

Comment:
I used to think the same until I read a UX article about this.

It's easy to know if a username is valid or not by attempting to register with that username, so by returning `Invalid username or password` we're not really making it any more secure, we're just making the UX worse.